// @turbopuffer/turbopuffer/internal/custom/respond-async.js
"use strict";
/**
 * Transparent polling for tpuf APIs that accept `prefer: respond-async`.
 *
 * Every API request is stamped with `prefer: respond-async`. If the server
 * applies the preference (i.e. responds with `202 Accepted` +
 * `preference-applied: respond-async`) the SDK polls the operation URL
 * until it finishes and returns the final result as if the call had been
 * synchronous.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
exports.prepareOptions = prepareOptions;
exports.maybePoll = maybePoll;
const tslib_1 = require("../tslib.js");
const Errors = tslib_1.__importStar(require("../../core/error.js"));
const headers_1 = require("../headers.js");
const sleep_1 = require("../utils/sleep.js");
const HEADER_PREFER = 'prefer';
const HEADER_PREFERENCE_APPLIED = 'preference-applied';
const HEADER_LOCATION = 'location';
const RESPOND_ASYNC = 'respond-async';
/** tunables; mutable so tests can override */
exports.config = {
    pollIntervalMs: 1000,
    pollRequestTimeoutMs: 60000,
};
function prepareOptions(options) {
    const normalized = (0, headers_1.buildHeaders)([options.headers]);
    // Don't overwrite an existing `prefer` header. This allows the caller to disable async mode.
    if (normalized.values.has(HEADER_PREFER) || normalized.nulls.has(HEADER_PREFER)) {
        return;
    }
    normalized.values.set(HEADER_PREFER, RESPOND_ASYNC);
    options.headers = normalized;
}
async function maybePoll(client, response, requestURL, options, clock) {
    if (!respondAsyncApplied(response))
        return response;
    const location = extractLocation(response, requestURL);
    const timeout = new Timeout(options.timeout ?? client.timeout);
    while (true) {
        if (timeout.remainingMs() === 0) {
            throw new Errors.APIConnectionTimeoutError();
        }
        const poll = await client.get(location, {
            headers: { [HEADER_PREFER]: null },
            timeout: timeout.pollTimeoutMs(),
            signal: options.signal,
        });
        const resp = resolvePollResponse(poll, clock);
        if (resp)
            return resp;
        await (0, sleep_1.sleep)(timeout.sleepDurationMs());
    }
}
function extractLocation(response, origURL) {
    const rawLocation = response.headers.get(HEADER_LOCATION);
    if (!rawLocation) {
        throw new Errors.TurbopufferError("server returned async response without a 'Location' header");
    }
    // Resolve the Location against the original request URL.
    let resolved;
    try {
        resolved = new URL(rawLocation, origURL);
    }
    catch {
        throw new Errors.TurbopufferError(`malformed 'Location' header: ${rawLocation}`);
    }
    // Reject a Location pointing at a different origin, to prevent API key exfiltration.
    const requestOrigin = new URL(origURL).origin;
    if (resolved.origin !== requestOrigin) {
        throw new Errors.TurbopufferError(`'Location' origin does not match request origin: ${rawLocation}`);
    }
    return resolved.toString();
}
function resolvePollResponse(body, clock) {
    if (body.status === 'running')
        return null;
    if (body.result && 'success' in body.result) {
        const resp = new Response(JSON.stringify(body.result.success), {
            status: 200,
            headers: { 'content-type': 'application/json' },
        });
        // Carry the original request's clock so client perf metrics propagate correctly.
        Object.defineProperty(resp, 'clock', { value: clock, enumerable: true });
        return resp;
    }
    if (body.result?.error) {
        const err = body.result.error;
        throw Errors.APIError.generate(err.status_code, err.detail, undefined, new Headers());
    }
    throw new Errors.TurbopufferError('malformed poll response');
}
function respondAsyncApplied(response) {
    if (response.status !== 202)
        return false;
    const applied = response.headers.get(HEADER_PREFERENCE_APPLIED) ?? '';
    return applied.trim().toLowerCase() === RESPOND_ASYNC;
}
const nowMs = () => Math.floor(performance.now());
class Timeout {
    constructor(timeoutMs) {
        this.deadline = timeoutMs === undefined ? Infinity : nowMs() + timeoutMs;
    }
    remainingMs() {
        return Math.max(this.deadline - nowMs(), 0);
    }
    pollTimeoutMs() {
        return Math.min(this.remainingMs(), exports.config.pollRequestTimeoutMs);
    }
    sleepDurationMs() {
        return Math.min(this.remainingMs(), exports.config.pollIntervalMs);
    }
}
