// @turbopuffer/turbopuffer/resources/namespaces.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Namespace = void 0;
const resource_1 = require("../core/resource.js");
const path_1 = require("../internal/utils/path.js");
const error_1 = require("../error.js");
class Namespace extends resource_1.APIResource {
    /**
     * Creates an instant, copy-on-write clone of a namespace.
     */
    branchFrom(params, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params;
        return this._client.post((0, path_1.path) `/v2/namespaces/${namespace}?stainless_overload=branchFrom`, {
            body: { branch_from_namespace: body },
            ...options,
        });
    }
    /**
     * Copy all documents from another namespace into this one.
     */
    copyFrom(params, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params;
        return this._client.post((0, path_1.path) `/v2/namespaces/${namespace}?stainless_overload=copyFrom`, {
            body: { copy_from_namespace: body },
            ...options,
        });
    }
    /**
     * Delete namespace.
     */
    deleteAll(params = {}, options) {
        const { namespace = this._client.defaultNamespace } = params ?? {};
        return this._client.delete((0, path_1.path) `/v2/namespaces/${namespace}`, options);
    }
    /**
     * Explain a query plan.
     */
    explainQuery(params = {}, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params ?? {};
        return this._client.post((0, path_1.path) `/v2/namespaces/${namespace}/explain_query`, { body, ...options });
    }
    /**
     * Signal turbopuffer to prepare for low-latency requests.
     */
    hintCacheWarm(params = {}, options) {
        const { namespace = this._client.defaultNamespace } = params ?? {};
        return this._client.get((0, path_1.path) `/v1/namespaces/${namespace}/hint_cache_warm`, options);
    }
    /**
     * Get metadata about a namespace.
     */
    metadata(params = {}, options) {
        const { namespace = this._client.defaultNamespace } = params ?? {};
        return this._client.get((0, path_1.path) `/v2/namespaces/${namespace}/metadata`, options);
    }
    /**
     * Issue multiple concurrent queries filter or search documents.
     */
    multiQuery(params, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params;
        return this._client.post((0, path_1.path) `/v2/namespaces/${namespace}/query?stainless_overload=multiQuery`, {
            body,
            ...options,
        });
    }
    /**
     * Query, filter, full-text search and vector search documents.
     */
    query(params = {}, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params ?? {};
        return this._client.post((0, path_1.path) `/v2/namespaces/${namespace}/query`, { body, ...options });
    }
    /**
     * Evaluate recall.
     */
    recall(params = {}, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params ?? {};
        return this._client.post((0, path_1.path) `/v1/namespaces/${namespace}/_debug/recall`, { body, ...options });
    }
    /**
     * Get namespace schema.
     */
    schema(params = {}, options) {
        const { namespace = this._client.defaultNamespace } = params ?? {};
        return this._client.get((0, path_1.path) `/v1/namespaces/${namespace}/schema`, options);
    }
    /**
     * Check whether the namespace exists.
     */
    async exists() {
        try {
            await this.schema();
            return true;
        }
        catch (e) {
            if (e instanceof error_1.NotFoundError) {
                return false;
            }
            throw e;
        }
    }
    /**
     * Update metadata configuration for a namespace.
     */
    updateMetadata(params = {}, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params ?? {};
        return this._client.patch((0, path_1.path) `/v1/namespaces/${namespace}/metadata`, { body, ...options });
    }
    /**
     * Update namespace schema.
     */
    updateSchema(params = undefined, options) {
        const { namespace = this._client.defaultNamespace, schema } = params ?? {};
        return this._client.post((0, path_1.path) `/v1/namespaces/${namespace}/schema`, { body: schema, ...options });
    }
    /**
     * Create, update, or delete documents.
     */
    write(params = {}, options) {
        const { namespace = this._client.defaultNamespace, ...body } = params ?? {};
        return this._client.post((0, path_1.path) `/v2/namespaces/${namespace}`, { body, maxRetries: 6, ...options });
    }
}
exports.Namespace = Namespace;
